src/account_manager/builder.rs

```diff
@@ -130,90 +129,89 @@ impl AccountManagerBuilder {
-                        .collect(),
-                )),
-                #[cfg(feature = "events")]
-                accounts: Arc::new(RwLock::new(
-                    accounts
-                        .into_iter()
-                        .map(|a| {
-                            AccountHandle::new(
-                                a,
-                                client.clone(),
-                                secret_manager.clone(),
-                                event_emitter.clone(),
-                                storage_manager.clone(),
-                            )
-                        })
-                        .collect(),
-                )),
-                background_syncing_status: Arc::new(AtomicUsize::new(0)),
-                client_options: Arc::new(RwLock::new(client_options)),
-                coin_type: Arc::new(AtomicU32::new(coin_type)),
-                secret_manager,
-                #[cfg(feature = "events")]
-                event_emitter,
-                storage_options,
-                storage_manager,
-            });
+        if self.coin_type.is_none() {
+            let coin_type = read_manager_builder
+                .and_then(|data| data.coin_type)
+                .ok_or(crate::Error::MissingParameter("coin_type (IOTA: 4218, Shimmer: 4219)"))?;
+
+            // Update self so it gets used and stored again
+            self.coin_type.replace(coin_type);
+        }
+
+        #[cfg(feature = "storage")]
+        // Store account manager data in storage
+        storage_manager.lock().await.save_account_manager_data(&self).await?;
+
+        let client = self
+            .client_options
+            .clone()
+            .ok_or(crate::Error::MissingParameter("client_options"))?
+            .finish()?;
+
+        #[cfg(feature = "events")]
+        let event_emitter = Arc::new(Mutex::new(EventEmitter::new()));
+
+        #[cfg(feature = "storage")]
+        let accounts = storage_manager.lock().await.get_accounts().await.unwrap_or_default();
+        #[cfg(not(feature = "storage"))]
+        let accounts = Vec::new();
+        let mut account_handles: Vec<AccountHandle> = accounts
+            .into_iter()
+            .map(|a| {
+                AccountHandle::new(
+                    a,
+                    client.clone(),
+                    self.secret_manager
+                        .clone()
+                        .expect("secret_manager needs to be provided"),
+                    #[cfg(feature = "events")]
+                    event_emitter.clone(),
+                    #[cfg(feature = "storage")]
+                    storage_manager.clone(),
+                )
+            })
+            .collect::<_>();
+
+        // If the manager builder is not set, it means the user provided it and we need to update the addresses.
+        // In the other case it was loaded from the database and addresses are up to date.
```

Comment:
```suggestion
        // If they are new, we need to update the accounts because the bech32 HRP might have changed.
        // In the other case it was loaded from the database and addresses are up to date.
```